core/arch/arm/plat-stm32mp1/plat_huk.c

```diff
@@ -0,0 +1,138 @@
+				      uint8_t *out, size_t *out_len)
+{
+	uint8_t uid[HW_UNIQUE_KEY_LENGTH] = { 0 };
+	uint8_t nonce[12] = {
+		0xd2, 0x45, 0x0e, 0x07, 0xea, 0x5d,
+		0xe0, 0x42, 0x6c, 0x0f, 0xa1, 0x33,
+	};
+	uint8_t tag[16] = { 0 };
+	TEE_Result res = TEE_ERROR_GENERIC;
+	size_t tag_len = sizeof(tag);
+	size_t uid_len = sizeof(uid);
+	uint32_t word = 0;
+	void *ctx = NULL;
+	size_t n = 0;
+
+	for (n = 0; n < (UID2_OTP - UID0_OTP + 1); n++) {
+		res = stm32_bsec_read_otp(&word, UID0_OTP + n);
+		if (res)
+			return res;
+
+		memcpy(uid + n * sizeof(uint32_t), &word, sizeof(uint32_t));
+	}
+
+	/* HUK is the AES-GCM encrypted UID using a secret key from OTP */
+	res = crypto_authenc_alloc_ctx(&ctx, TEE_ALG_AES_GCM);
+	if (res)
+		return res;
+
+	res = crypto_authenc_init(ctx, TEE_MODE_ENCRYPT, key, key_len,
+				  nonce, sizeof(nonce),
+				  TEE_AES_BLOCK_SIZE, 0, uid_len);
+	if (res)
+		return res;
+
+	res = crypto_authenc_enc_final(ctx, uid, sizeof(uid), out, out_len,
+				       tag, &tag_len);
+	if (res)
+		return res;
+
+	crypto_authenc_final(ctx);
+	crypto_authenc_free_ctx(ctx);
+
+	return TEE_SUCCESS;
+}
+
+TEE_Result tee_otp_get_hw_unique_key(struct tee_hw_unique_key *hwkey)
+{
+	uint32_t bsec_base = CFG_STM32MP1_HUK_BSEC_BASE;
+	uint8_t otp_data[HW_UNIQUE_KEY_LENGTH] = { 0 };
+	TEE_Result res = TEE_ERROR_GENERIC;
+	size_t len = HW_UNIQUE_KEY_LENGTH;
+	uint32_t word = 0;
+	uint32_t tmp = 0;
+	bool lock = false;
+	size_t n = 0;
+	static bool initialized;
+	const char *state = "Locked";
+	const char *type = "OTP";
+
+	/* Read 128 bits of OTP data (the secret) */
+	for (n = 0; n < HW_UNIQUE_KEY_LENGTH / sizeof(uint32_t); n++) {
+		res = stm32_bsec_read_permanent_lock(bsec_base + n, &lock);
+		if (res) {
+			EMSG("huk: read permanent lock");
+			return res;
+		}
+
+		if (!lock)
+			state = "Devel";
+
+		res = stm32_bsec_read_otp(&tmp, bsec_base + n);
+		if (res)
+			return res;
+
+		res = stm32_bsec_shadow_register(bsec_base + n);
+		if (res) {
+			EMSG("huk: access shadow register");
+			return res;
+		}
+
+		res = stm32_bsec_read_otp(&word, bsec_base + n);
+		if (res) {
+			EMSG("huk: read OTP secret");
+			return res;
+		}
+
+		res = stm32_bsec_write_otp(tmp, bsec_base + n);
+		if (res) {
+			EMSG("huk: shadow the secret");
+			return res;
+		}
+
+		memcpy(otp_data + n * sizeof(uint32_t), &word,
+		       sizeof(uint32_t));
+	}
+
+	if (IS_ENABLED(CFG_STM32MP1_HUK_BSEC_DERIVE_GCM)) {
+		/* The secret is not per-device unique: derive using the UID */
+		type = "Derived";
+		res = aes_gcm_encrypt_uid(otp_data, HW_UNIQUE_KEY_LENGTH,
+					  hwkey->data,  &len);
+		if (res || len != HW_UNIQUE_KEY_LENGTH) {
+			EMSG("huk: derive failure");
+			return TEE_ERROR_GENERIC;
+		}
+	} else {
+		/* The secret is unique: use it as the HUK */
+		memcpy(hwkey->data, otp_data, HW_UNIQUE_KEY_LENGTH);
+	}
+
+	if (!initialized) {
+		initialized = true;
+		IMSG("HUK [%s %s]", type, state);
+		DHEXDUMP(hwkey->data, HW_UNIQUE_KEY_LENGTH);
+	}
+
```

Comment:
It would be nice to have `memzero_explicit(otp_data, sizeof(otp_data))` + `memzero_explicit(&tmp, sizeof(tmp))` on exit (including some error paths)?